src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -596,106 +924,289 @@ namespace nbl::ext::imgui
-				memReq.memoryTypeBits &= m_device->getPhysicalDevice()->getUpStreamingMemoryTypeBits();
-				auto memOffset = m_device->allocate(memReq, vertexBuffer.get());
-				assert(memOffset.isValid());
-			}
+				core::vector2df_SIMD toNDC(core::vector2df_SIMD in) const
+				{
+					return in * scale + translate;
+				}
+			};
 
-			IGPUBuffer::SCreationParams indexCreationParams = {};
-			indexCreationParams.usage = nbl::core::bitflag(nbl::asset::IBuffer::EUF_VERTEX_BUFFER_BIT) | nbl::asset::IBuffer::EUF_INDEX_BUFFER_BIT
-				| nbl::asset::IBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF;
-			indexCreationParams.size = indexSize;
+			const TRS trs = [&]() 
+			{
+				TRS retV;
 
-			auto & indexBuffer = m_indexBuffers[frameIndex];
+				retV.scale = core::vector2df_SIMD{ 2.0f / drawData->DisplaySize.x , 2.0f / drawData->DisplaySize.y };
+				retV.translate = core::vector2df_SIMD { -1.0f, -1.0f } - core::vector2df_SIMD{ drawData->DisplayPos.x, drawData->DisplayPos.y } * trs.scale;
 
-			if (static_cast<bool>(indexBuffer) == false || indexBuffer->getSize() < indexSize)
-			{
-				indexBuffer = m_device->createBuffer(std::move(indexCreationParams));
+				return std::move(retV);
+			}();
 
-				video::IDeviceMemoryBacked::SDeviceMemoryRequirements memReq = indexBuffer->getMemoryReqs();
-				memReq.memoryTypeBits &= m_device->getPhysicalDevice()->getUpStreamingMemoryTypeBits();
-				auto memOffset = m_device->allocate(memReq, indexBuffer.get());
-				assert(memOffset.isValid());
-			}
+			struct MDI_PARAMS
+			{
+				std::array<MDI_SIZE_TYPE, MDI_COMPONENT_COUNT> bytesToFill = {}; //! used with MDI::E_BUFFER_CONTENT for elements
+				MDI_SIZE_TYPE totalByteSizeRequest = {}, //! sum of bytesToFill
+				drawCount = {};	//! amount of objects to draw for indirect call request
+			};
 
+			const MDI_PARAMS mdiParams = [&]()
 			{
-				auto vBinding = vertexBuffer->getBoundMemory();
-				auto iBinding = indexBuffer->getBoundMemory();
+				MDI_PARAMS params;
 
+				for (uint32_t i = 0; i < drawData->CmdListsCount; i++)
 				{
-					if (!vBinding.memory->map({ 0ull, vBinding.memory->getAllocationSize() }, IDeviceMemoryAllocation::EMCAF_READ))
-						logger->log("Could not map device memory for vertex buffer data!", system::ILogger::ELL_WARNING);
-
-					assert(vBinding.memory->isCurrentlyMapped());
+					const ImDrawList* commandList = drawData->CmdLists[i];
+					params.bytesToFill[MDI::EBC_DRAW_INDIRECT_STRUCTURES] += commandList->CmdBuffer.Size * sizeof(VkDrawIndexedIndirectCommand);
+					params.bytesToFill[MDI::EBC_ELEMENT_STRUCTURES] += commandList->CmdBuffer.Size * sizeof(PerObjectData);
 				}
+				params.bytesToFill[MDI::EBC_VERTEX_BUFFERS] = drawData->TotalVtxCount * sizeof(ImDrawVert);
+				params.bytesToFill[MDI::EBC_INDEX_BUFFERS] = drawData->TotalIdxCount * sizeof(ImDrawIdx);
+
+				// calculate upper bound byte size limit for mdi buffer
+				params.totalByteSizeRequest = std::reduce(std::begin(params.bytesToFill), std::end(params.bytesToFill));
```

Comment:
you need to pessimize the size needed due to alignments, so instead of doing a reduction, use a make-believe/dry-run linear allocator over a very large address space (2GB) and query the `used_space`/cursor position

Reply:
done by myself